bob-access/src/credentials.rs

```diff
@@ -0,0 +1,159 @@
+    Token(String),
+}
+
+#[derive(Clone, Copy, PartialEq)]
+pub enum CredentialsType {
+    Stub,
+    Basic,
+    Token,
+}
+
+impl CredentialsType {
+    pub fn is_basic(&self) -> bool {
+        *self == CredentialsType::Basic
+    }
+
+    pub fn is_stub(&self) -> bool {
+        *self == CredentialsType::Stub
+    }
+}
+
+fn _credentials_type(tp: Option<CredentialsType>) -> CredentialsType {
+    static mut CREDENTIALS_TYPE: CredentialsType = CredentialsType::Stub;
+    unsafe {
+        if let Some(tp) = tp {
+            CREDENTIALS_TYPE = tp;
+        }
+        CREDENTIALS_TYPE
+    }
+}
+
+pub fn credentials_type() -> CredentialsType {
+    _credentials_type(None)
+}
+
+pub fn set_credentials_type(tp: CredentialsType) {
+    _credentials_type(Some(tp));
+}
+
+impl Credentials {
+    pub fn builder() -> CredentialsBuilder {
+        CredentialsBuilder::default()
+    }
+
+    pub fn username(&self) -> Option<&str> {
+        let kind = self.kind.as_ref()?;
+        match kind {
+            CredentialsKind::Basic {
+                username,
+                password: _,
+            } => Some(username),
+            CredentialsKind::Token(_) => None,
+        }
+    }
+
+    pub fn password(&self) -> Option<&str> {
+        let kind = self.kind.as_ref()?;
+        match kind {
+            CredentialsKind::Basic {
+                username: _,
+                password,
+            } => Some(password),
+            CredentialsKind::Token(_) => None,
+        }
+    }
+
+    pub fn ip(&self) -> Option<IpAddr> {
+        Some(self.address?.ip())
+    }
+
+    pub fn is_complete(&self) -> bool {
+        self.address.is_some() && self.ip().is_some()
+    }
+}
+
+#[derive(Debug, Default)]
+pub struct CredentialsBuilder {
+    kind: Option<CredentialsKind>,
+    address: Option<SocketAddr>,
+}
+
+impl CredentialsBuilder {
+    pub fn with_username_password(
+        &mut self,
+        username: impl Into<String>,
+        password: impl Into<String>,
+    ) -> &mut Self {
+        self.kind = Some(CredentialsKind::Basic {
+            username: username.into(),
+            password: password.into(),
+        });
+        self
+    }
+
+    pub fn with_token(&mut self, token: impl Into<String>) -> &mut Self {
+        self.kind = Some(CredentialsKind::Token(token.into()));
+        self
+    }
+
+    pub fn with_address(&mut self, address: Option<SocketAddr>) -> &mut Self {
+        self.address = address;
+        self
+    }
+
+    pub fn build(&self) -> Credentials {
+        Credentials {
+            address: self.address,
+            kind: self.kind.clone(),
+        }
+    }
+}
+
+#[async_trait]
+impl<B> FromRequest<B> for Credentials
+where
+    B: Send,
+{
+    type Rejection = Error;
+
+    async fn from_request(req: &mut RequestParts<B>) -> Result<Self, Self::Rejection> {
+        req.extract()
+    }
+}
+
+impl<T> From<&Request<T>> for Credentials {
+    fn from(req: &Request<T>) -> Self {
+        let md = req.metadata();
```

Comment:
This function is incomplete. It does not have `token` extraction. Obviously, it should somehow use `extract` function from `ExtractorExt` implementation